File: src/main/scala/org/deeplearning4j/scalnet/layers/core/Node.scala
/*
 * Copyright 2016 Skymind
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.deeplearning4j.scalnet.layers.core

/**
  * Trait for node in DL4J neural networks and computational graphs.
  * Nodes are assumed to have inputs and outputs with "shapes."
  *
  * @author David Kale
  */
trait Node {

  def name: String

  def inputShape: List[Int]

  def outputShape: List[Int]

  def reshapeInput(nIn: List[Int]): Node = this

  def describe(): String = "in=" + inputShape + " out=" + outputShape

}

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/execution/Node.java
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.nd4j.autodiff.execution;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.List;

/**
 * Intermediate Node representation
 *
 * @author raver119@gmail.com
 */
@Data
@Slf4j
@NoArgsConstructor
@ToString(exclude = {"opExecAction"})
public class Node {
    private int id;
    private String name;
    private List<Integer> input = new ArrayList<>();
    private List<Integer> output = new ArrayList<>();
    private List<Integer> unresolved = new ArrayList<>();
    private int[] originalOutput;
}
